src/text.c

```diff
@@ -0,0 +1,410 @@
+	}
+	FLUSH;
+}
+
+static void write_value(FILE *f, const char *v) {
+	// write and escape value
+	const char *p = v;
+	const char *n = p;
+	while (*n) {
+		switch (*n) {
+		case '\\':
+			SKIP;
+			fwrite ("\\\\", 1, 2, f);
+			break;
+		case '\n':
+			SKIP;
+			fwrite ("\\n", 1, 2, f);
+			break;
+		}
+		n++;
+	}
+	FLUSH;
+}
+
+static bool text_fsave(Sdb *s, FILE *f, bool sort, SdbList *path) {
+	//path
+	write_path (f, path);
+	fwrite ("\n", 1, 1, f);
+
+	// k=v entries
+	SdbList *l = sdb_foreach_list (s, sort);
+	SdbKv *kv;
+	SdbListIter *it;
+	ls_foreach (l, it, kv) {
+		const char *k = sdbkv_key (kv);
+		const char *v = sdbkv_value (kv);
+		// escape leading '/'
+		if (*k == '/') {
+			fwrite ("\\", 1, 1, f);
+		}
+
+		write_key (f, k);
+		fwrite ("=", 1, 1, f);
+		write_value (f, v);
+
+		fwrite ("\n", 1, 1, f);
+	}
+	ls_free (l);
+
+	// sub-namespaces
+	l = s->ns;
+	if (sort) {
+		l = ls_clone (l);
+		ls_sort (l, cmp_ns);
+	}
+	SdbNs *ns;
+	ls_foreach (l, it, ns) {
+		fwrite ("\n", 1, 1, f);
+		ls_push (path, ns->name);
+		text_fsave (ns->sdb, f, sort, path);
+		ls_pop (path);
+	}
+	if (l != s->ns) {
+		ls_free (l);
+	}
+
+	return true;
+}
+#undef FLUSH
+#undef SKIP
+
+SDB_API bool sdb_text_fsave(Sdb *s, FILE *f, bool sort) {
+	SdbList *path = ls_new ();
+	if (!path) {
+		return false;
+	}
+	bool r = text_fsave (s, f, sort, path);
+	ls_free (path);
+	return r;
+}
+
+SDB_API bool sdb_text_save(Sdb *s, const char *file, bool sort) {
+	FILE *f = fopen (file, "w");
+	if (!f) {
+		return false;
+	}
+	bool r = sdb_text_fsave (s, f, sort);
+	fclose (f);
+	return r;
+}
+
+typedef struct {
+	FILE *f;
+	bool eof;
+	size_t bufsz;
+	char *buf;
+	size_t buf_filled; // how many bytes of buf are currently filled with data from the file
+	Sdb *root_db;
+	Sdb *cur_db; // current namespace, changes when encountering a path line
+	// can't use pointers here because buf may be moved on realloc
+	size_t pos; // current processing position in the buffer
+	size_t token_begin; // beginning of the currently processed token in the buffer
+	size_t shift; // amount to shift chars to the left (from unescaping)
+	SdbList/*<size_t>*/ *path;
+	enum { STATE_NEWLINE, STATE_PATH, STATE_KEY, STATE_VALUE } state;
+	bool unescape; // whether the prev char was a backslash, i.e. the current one is escaped
+} LoadCtx;
+
+// load more data from ctx->f into ctx->buf starting at ctx->pos
+// and update ctx->buf_filled and ctx->eof.
+// may realloc and change ctx->buf and ctx->bufsz.
+static bool load_read_more(LoadCtx *ctx) {
+	// realloc if no space
+	if (ctx->pos >= ctx->bufsz - 1) {
+		size_t newsz = ctx->bufsz + ctx->bufsz;
+		if (newsz < ctx->bufsz) {
+			return false;
+		}
+		ctx->bufsz = newsz;
+		char *newbuf = realloc (ctx->buf, ctx->bufsz);
+		if (!newbuf) {
+			return false;
+		}
+		ctx->buf = newbuf;
+	}
+	// read full buffer (leave 1 for null-terminating)
+	size_t red = fread (ctx->buf + ctx->pos, 1, ctx->bufsz - 1 - ctx->pos, ctx->f);
+	if (!red && !feof (ctx->f)) {
```

Comment:
I'm not sure `feof` requires `red == 0`. At least the man page doesn't say so, so I guess there can be cases where it reads few bytes and set the status of the file to EOF.

Reply:
Yes, that is correct. `feof()` will most of the time be set after reading some bytes. But then in the next iteration, `red` will always be `0` and `feof()` still true so it will exit there. This just makes the code a bit simpler.

Reply:
i can confirm this behaviour despite not being documented

Reply:
It is documented, just phrased differently and in different places.